### PR TITLE
Have an option to specify custom apiserver service IP

### LINKED ISSUE #82743: Have option to pass in cluster service IP for apiserver
**What would you like to be added**: Have an option to pass in cluster service IP for the apiserver. Currently the first IP of the ServiceClusterIPRange is used as the cluster IP for the apiserver where ServiceClusterIPRange can be changed using a flag and it would be nice to have service IP also configurable using the flag. 

**Why is this needed**: This will help in setting up the network configuration prior to cluster creation when using custom container network plugins for in cluster routing.

### PR DESCRIPTION
**Pull request by kmala**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This will provide an option to override the apiserver cluster serviceIP. 
**Which issue(s) this PR fixes**:

Fixes #82743

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@kmala can you explain what's the difference between the --advertise-address and the ServiceClusterIPRange? I saw your comment in the issue said the ServiceClusterIPRange is "incluster service IP". Is the difference that one is in-cluster and the other one is external?

**Comment by kmala** [comment]
Yes, you are right `--advertise-address` is IP where the apiserver will be bind to while starting the server(external) and is usually the host IP if unspecified. ServiceClusterIP is the IP of the kubernetes service (of type ClusterIP) which apiserver will create during the bootstrap.

**Comment by caesarxuchao** [comment]
> "ServiceClusterIP is the IP of the kubernetes service (of type ClusterIP)"

Are you talking the "kubernetes" service in the "default" namespace? IIRC, the address in that service is populated from `--advertise-address`, not `ServiceClusterIP`.

**Comment by kmala** [comment]
yes, it is the kubernetes service in the default namespace which currently is the serviceIPRange base +1.

**Comment by kmala** [comment]
@caesarxuchao can you review this PR as this pending for some time?

**Comment by liggitt** [comment]
If I'm reading this correctly, this adds a configuration option without proving it actually works with tests. I see @thockin commented in the parent issue, but it would be helpful to know how common a use case this is. Are there parts of the apiserver that assume the kubernetes service IP is the first in the range? I'm wary of adding more untested configuration options to the apiserver without strong justification.

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -195,6 +198,10 @@ func (s *ServerRunOptions) Flags() (fss cliflag.NamedFlagSets) {
 		"of type NodePort, using this as the value of the port. If zero, the Kubernetes master "+
 		"service will be of type ClusterIP.")
 
+	fs.StringVar(&s.KubernetesServiceClusterIP, "kubernetes-service-cluster-ip", s.KubernetesServiceClusterIP, ""+
+		"If non-empty, the Kubernetes master service (which apiserver creates/maintains) will use "+
+		"the IP as cluster-internal ip. If empty, it will be service cluster ip range base + 1.")
```
Comment (seh) [inline cmd/kube-apiserver/app/options/options.go]: s/the IP as/the IP address as/
s/ip\./IP address./
s/be service cluster IP/
s/be the service cluster IP/

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -694,10 +677,18 @@ func getServiceIPAndRanges(serviceClusterIPRanges string) (net.IP, net.IPNet, ne
 	var primaryServiceIPRange net.IPNet
 	var secondaryServiceIPRange net.IPNet
 	var err error
+
+	if kubernetesServiceClusterIP != "" {
+		apiServerServiceIP = net.ParseIP(kubernetesServiceClusterIP)
+		if apiServerServiceIP == nil {
+			return net.IP{}, net.IPNet{}, net.IPNet{}, fmt.Errorf("kubernetes-service-cluster-ip %s is not a valid IP", kubernetesServiceClusterIP)
```
Comment (seh) [inline cmd/kube-apiserver/app/server.go]: s/valid IP/valid IP address/

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -724,6 +715,10 @@ func getServiceIPAndRanges(serviceClusterIPRanges string) (net.IP, net.IPNet, ne
 			return net.IP{}, net.IPNet{}, net.IPNet{}, fmt.Errorf("service-cluster-ip-range[1] is not an ip net")
 		}
 		secondaryServiceIPRange = *secondaryServiceClusterCIDR
+		secondaryServiceIPRange, _, err = master.ServiceIPRange(secondaryServiceIPRange, apiServerServiceIP)
+		if err != nil {
+			return net.IP{}, net.IPNet{}, net.IPNet{}, fmt.Errorf("error determining service IP ranges for secondary service cidr: %v", err)
```
Comment (seh) [inline cmd/kube-apiserver/app/server.go]: s/cidr/CIDR/

File: pkg/master/services.go [hunk 4]
```diff
@@ -43,12 +44,16 @@ func ServiceIPRange(passedServiceClusterIPRange net.IPNet) (net.IPNet, net.IP, e
...
-	if err != nil {
-		return net.IPNet{}, net.IP{}, err
+	if passedAPIServiceClusterIP == nil {
+		// Select the first valid IP from ServiceClusterIPRange to use as the GenericAPIServer service IP.
+		passedAPIServiceClusterIP, err = utilnet.GetIndexedIP(&serviceClusterIPRange, 1)
+		if err != nil {
+			return net.IPNet{}, net.IP{}, err
+		}
+	} else if !serviceClusterIPRange.Contains(passedAPIServiceClusterIP) {
+		return net.IPNet{}, net.IP{}, fmt.Errorf("the service cluster IP must be in the range of %s", serviceClusterIPRange)
```
Comment (seh) [inline pkg/master/services.go]: s/IP/IP address/

**Comment by lavalamp** [comment]
Forgive me if I missed something as I only did a quick skim, but this looks to be a footgun in two big ways right now:
* A cluster is totally broken if the apiserver flag ever changes, or if separate apiservers disagree on its value.
* Where does this value get consumed from the table such that it won't be assigned to other services?

File: cmd/kube-apiserver/app/options/options.go [hunk 5]
```diff
@@ -50,18 +50,21 @@ type ServerRunOptions struct {
...
+	KubernetesServiceNodePort  int
+	KubernetesServiceClusterIP string
+	MaxConnectionBytesPerSec   int64
 	// ServiceClusterIPRange is mapped to input provided by user
 	ServiceClusterIPRanges string
 	//PrimaryServiceClusterIPRange and SecondaryServiceClusterIPRange are the results
 	// of parsing ServiceClusterIPRange into actual values
 	PrimaryServiceClusterIPRange   net.IPNet
 	SecondaryServiceClusterIPRange net.IPNet
+	// APIServerServiceIP is the result of parsing KubernetesServiceClusterIP
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: This pair of names is confusing. `KubernetesServiceClusterIPString` and `KubernetesServiceClusterIP` are better, or use `...IP` and `...ParsedIP`, something like that. But don't call the same thing both a service IP and a cluster IP.

**Comment by kmala** [comment]
> * A cluster is totally broken if the apiserver flag ever changes, or if separate apiservers disagree on its value.

Not sure what you meant by cluster is totally broken but if the value of the flag is changed or if a value is specified to the flag then apiserver service will have that new IP and any service which connects to apiserver will face issue and have to reconnect using the new IP but the same is true even for the most flag.

> * Where does this value get consumed from the table such that it won't be assigned to other services?

There is repair controller and ipallocator module which will make sure that the same IP is not assigned to another service.

**Comment by kmala** [comment]
>  I see @thockin commented in the parent issue, but it would be helpful to know how common a use case this is. 

This is very important for us because we can't run kube-proxy because of IPVLAN networking and most/all additional tooling/software connect to apiserver using the cluster ip.

> Are there parts of the apiserver that assume the kubernetes service IP is the first in the range? 

I checked the codebase and couldn't find anywhere where it is assumed the first in the range.

**Comment by lavalamp** [comment]
>>A cluster is totally broken if the apiserver flag ever changes, or if separate apiservers disagree on its value.

>Not sure what you meant by cluster is totally broken but if the value of the flag is changed or if a value is specified to the flag then apiserver service will have that new IP and any service which connects to apiserver will face issue and have to reconnect using the new IP but the same is true even for the most flag.

What happens to other processes running in the cluster using the IP given via the downward API? They all stop working until restarted. `kubelet` sets this env var before running things, but does it expect this to ever change and refresh it? Probably not, but I don't know for sure. Most likely, changing this requires everything in the cluster to restart simultaneously.

Also what happens if multiple apiservers disagree over the value here (e.g., in the middle of a rolling restart)?

>>Where does this value get consumed from the table such that it won't be assigned to other services?

>There is repair controller and ipallocator module which will make sure that the same IP is not assigned to another service.

I don't see that code touched in this PR though. Where does it do that? What if some other service is already using the value specified? If this changes, where does it release the prior value?

**Comment by kmala** [comment]
> > > A cluster is totally broken if the apiserver flag ever changes, or if separate apiservers disagree on its value.
> 
> > Not sure what you meant by cluster is totally broken but if the value of the flag is changed or if a value is specified to the flag then apiserver service will have that new IP and any service which connects to apiserver will face issue and have to reconnect using the new IP but the same is true even for the most flag.
> 
> What happens to other processes running in the cluster using the IP given via the downward API? They all stop working until restarted. `kubelet` sets this env var before running things, but does it expect this to ever change and refresh it? Probably not, but I don't know for sure. Most likely, changing this requires everything in the cluster to restart simultaneously.
> 
> Also what happens if multiple apiservers disagree over the value here (e.g., in the middle of a rolling restart)?

I agree that if you need to change the value in a running cluster, then all the clients of the apiserver who are using the cluster IP needs to restart but that is something cluster operator knows when changing the apiserver config since the default value is the existing behavior. The way i see this flag is as an additional knob for the users who can use it if they need thus making kubernetes more extensible. Also, there are other flag for apiserver where clients needs to restart on change in the value( ex.service-cluster-ip-range).

> > > Where does this value get consumed from the table such that it won't be assigned to other services?
> 
> > There is repair controller and ipallocator module which will make sure that the same IP is not assigned to another service.
> 
> I don't see that code touched in this PR though. Where does it do that? What if some other service is already using the value specified? If this changes, where does it release the prior value?

The [repair controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/controller/repair.go) will update the currently used service IP's and this will make sure that any new service getting created will make sure that it is an usused IP. If the IP is already used which means when someone is changing an existing cluster, the kubernetes service will not be created, so the user has to start apiserver with a different IP.

**Comment by lavalamp** [comment]
OK so the service is created [here](https://github.com/kubernetes/kubernetes/blob/4ee00204ae2a7ce631674ededc1745a0341bce71/pkg/master/controller.go#L239) and we rely on the regular system for accounting for it. That's better than I thought it might be, but still is problematic when:
* apiservers run with skewed flags
* some other service has chosen the IP already

Basically, all problems you get when this changes.

I think this feature needs to somehow make this immutable and prevent it from changing, there are too many problems to overcome to make that safe. At a bare minimum, the flag help string and other documentation needs to explain that changing it for an already existing cluster is very dangerous. But I'd much prefer if we noticed a change and refused to start. I agree we have other similarly dangerous flags that aren't clearly marked and should be--that's not a reason to add more :)

Also even running the first time, I believe it is actually technically possible for someone to connect to apiserver and construct a service before the kubernetes service is constructed. Today we are lucky (looks like there's only 100ms for this race to happen) and/or people are trained not to use the first IP in the range.

A note on the code here since I had to look through this a bit more to dig up where the service is created: it seems for several functions which were returning this, you added a parameter so they know what to return? I'd much prefer removing the return parameter instead.

**Comment by kmala** [comment]
> OK so the service is created [here](https://github.com/kubernetes/kubernetes/blob/4ee00204ae2a7ce631674ededc1745a0341bce71/pkg/master/controller.go#L239) and we rely on the regular system for accounting for it. That's better than I thought it might be, but still is problematic when:
> 
> * apiservers run with skewed flags
> * some other service has chosen the IP already

Can you elaborate on the skewed flags issue? For the issue where other service already chosen the IP, currently the apiserver service will not be created but if you think a different behavior is better, we can discuss and i can implement it.

> Basically, all problems you get when this changes.
> 
> I think this feature needs to somehow make this immutable and prevent it from changing, there are too many problems to overcome to make that safe. At a bare minimum, the flag help string and other documentation needs to explain that changing it for an already existing cluster is very dangerous. But I'd much prefer if we noticed a change and refused to start. I agree we have other similarly dangerous flags that aren't clearly marked and should be--that's not a reason to add more :)

I agree that we should have all the proper precautions to make sure users are aware of the effects of using this flag and ready to make changes to that effect. If the cluster operator is making the change to config then i don't think we should prevent it as he might be well aware of the effects of the change.

> Also even running the first time, I believe it is actually technically possible for someone to connect to apiserver and construct a service before the kubernetes service is constructed. Today we are lucky (looks like there's only 100ms for this race to happen) and/or people are trained not to use the first IP in the range.
> 
> A note on the code here since I had to look through this a bit more to dig up where the service is created: it seems for several functions which were returning this, you added a parameter so they know what to return? I'd much prefer removing the return parameter instead.

i am assuming you are talking about `getServiceIPAndRanges` function, otherwise can you comment on the code so that i can make the change.

**Pull request closed without merging**